hax/hax/motr/delivery.py

```diff
@@ -124,3 +123,24 @@ def notify_delivered(self, message_id: MessageId):
                     with client:
                         client.notify()
                     return
+            # If notify_delivered() was invoked before wait_for_all(), i.e.
+            # if the ha message is already delivered before the sender starts
+            # waiting on the same, append the delivered message to the list of
+            # recently_delivered, so that the delivery is found before
+            # wait_for_{all, any}() starts the conditional wait.
+            self.recently_delivered[HaLinkMessagePromise(
```

Comment:
This change will most probably cause `recently_delivered` dictionary to grow unconditionally.

The fact is that if a message is issued from our `hax.c` and nobody invokes `wait_for_all()`/`wait_for_any()`, the corresponding promise will never be removed from this dictionary.

I understand that the problem you're solving is real and we need this kind of fix. But we also need a mechanism that will remove these orphan promises from this dict eventually.

Options:
1. Try [weak references](https://docs.python.org/3/library/weakref.html#module-weakref) to get promises garbage-collected eventually.
   - Cons: this removal is a kind of uncontrolled. We don't know exactly when a key or value will be removed. So there is even a chance that a "good" (not orphaned, i.e. the one being awaited by somebody) will be removed too early.
2. Manual TTL mechanism with a background thread that traverses the keys and checks if they seem to be too old.

Reply:
Update: the corresponding change is now part of this PR. @mssawant please review once you have time.